wordninja.py

```diff
@@ -2,40 +2,50 @@
 from math import log
 
 
-
-__version__ = '0.1.4'
+__version__ = '0.1.5'
 
 
 # I did not author this code, only tweaked it from:
 # http://stackoverflow.com/a/11642687/2449774
 # Thanks Generic Human!
 
 
+# Modifications by Scott Randal (Genesys)
+#
+# 1. Preserve original character case after splitting
+# 2. Avoid splitting every post-digit character in a mixed string (e.g. 'win32intel')
+# 3. Avoid splitting digit sequences
+# 4. Handle input containing apostrophes (for possessives and contractions)
+#
+# Dictionary changes:
+# Change 2 required adding single digits to the dictionary
+# Change 4 required the following dictionary additions:
+#   's
+#   '
+#   <list of contractions>
+
 
 # Build a cost dictionary, assuming Zipf's law and cost = -math.log(probability).
 with gzip.open(os.path.join(os.path.dirname(os.path.abspath(__file__)),'wordninja','wordninja_words.txt.gz')) as f:
   words = f.read().decode().split()
 _wordcost = dict((k, log((i+1)*log(len(words)))) for i,k in enumerate(words))
 _maxword = max(len(x) for x in words)
-_SPLIT_RE = re.compile("[^a-zA-Z0-9]+")
-
+_SPLIT_RE = re.compile("[^a-zA-Z0-9']+")
 
 
 def split(s):
   """Uses dynamic programming to infer the location of spaces in a string without spaces."""
-  s = s.lower()
   l = [_split(x) for x in _SPLIT_RE.split(s)]
   return [item for sublist in l for item in sublist]
 
 
-
 def _split(s):
     # Find the best match for the i first characters, assuming cost has
     # been built for the i-1 first characters.
     # Returns a pair (match_cost, match_length).
     def best_match(i):
         candidates = enumerate(reversed(cost[max(0, i-_maxword):i]))
-        return min((c + _wordcost.get(s[i-k-1:i], 9e999), k+1) for k,c in candidates)
+        return min((c + _wordcost.get(s[i-k-1:i].lower(), 9e999), k+1) for k,c in candidates)
```

Comment:
so you're saving the .lower() operation till after the split--is that because you wanted to split camel-case items?

Reply:
Doing it here allows the wordlist lookup without forcing the final output to lower case. So if the input were camel-case or all-caps, that would be preserved in the output.